bucket/ant.json

```
@@ -9,6 +9,11 @@
     "url": "https://www.apache.org/dist/ant/binaries/apache-ant-1.10.12-bin.zip",
     "hash": "sha512:7e6fbcc3563df4bd87c883ad86a161a71da2774e0ed71a1b3aad82cbff1a7656ed9a0acb5ce40652129376dfd79f1ef74ec3369c1067d412a63062fea62ceccd",
     "extract_dir": "apache-ant-1.10.12",
+    "pre_install": [
+        "if (Test-Path -PathType Container \"$persist_dir\\lib\") {",
+        "    Copy-Item \"$persist_dir\\lib\\*\" \"$dir\\lib\"",
+        "}"
+    ],
```

```suggestion
    ],
    "uninstaller": {
        "script": [
            "ensure \"$persist_dir\\lib\" | Out-Null",
            "Copy-Item \"$dir\\lib\\*\" \"$persist_dir\\lib\""
        ]
    },
```

You'll also need to copy the files to `$persist_dir` when uninstalling.

---

I'm not sure we want to do this. Ant has most of its version specific implementations in `$dir/lib`. This would not be copying out only user-specified libs, but _all_ libs.

---

Oh. I see, 
but my concern is that many users will just put their libs at `$dir\libs` instead of `$persist_dir\libs`, and see their libs disappear after updating the app.

A possible way to solve this is by putting **a list of built-in libs** in a JSON file, then, avoid copying them in `uninstaller`.

```
    "pre_install": [
        "# Copy user libs to $dir",
        "if (Test-Path \"$persist_dir\\lib\") {",
        "    Get-ChildItem \"$persist_dir\\lib\" | Select -ExpandProperty Name | ForEach-Object {",
        "        if (!(Test-Path \"$dir\\lib\\$_\")) { Copy-Item \"$persist_dir\\lib\\$_\" \"$dir\\lib\" }",
        "    }",
        "}",
        "# Record built-in libs in builtin_libs.json",
        "$builtin_libs = Get-ChildItem \"$dir\\lib\" | Select -ExpandProperty Name",
        "$builtin_libs | ConvertTo-Json | Out-File \"$dir\\builtin_libs.json\" -Encoding Ascii"
    ],
    "uninstaller": {
        "script": [
            "# Only persist libs that are added by the user, but not built-in ones",
            "ensure \"$persist_dir\\lib\" | Out-Null",
            "$builtin_libs = Get-Content \"$dir\\builtin_libs.json\" | ConvertFrom-Json",
            "Get-ChildItem \"$dir\\lib\" | Select -ExpandProperty Name | ForEach-Object {",
            "    if (!($builtin_libs -contains \"$_\")) { Copy-Item \"$dir\\lib\\$_\" \"$persist_dir\\lib\" }",
            "}"
        ]
    },
```

*p.s. The `Select -ExpandProperty Name` mainly exists to avoid errors when there are *spaces* in SCOOP_DIR.*

---

This works. I'll implement that. However, I'll record the list of built-in libs before copying the user-defined ones, instead of the other way around.

---

I'll leave the `-ExpandProperty` as suggested, but would the fact that we put quotes around the file names when we refer to them avoid the issue with spaces in the path?

---

Sorry that I didn't say it clearly. I tried to use `$_` in the script (which is a *DirectoryInfo* object), but error occur when I test it on a VM with Scoop directory that contains spaces.